### PR TITLE
Add `title` filters

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of https://github.com/djc/askama/issues/18.

I also used this opportunity to update the list in the first message of the issue.

### CODE REVIEW

File: askama/src/filters/mod.rs [hunk 1]
```diff
@@ -511,6 +511,39 @@ pub fn wordcount(s: impl ToString) -> Result<usize, Infallible> {
...
+    let mut need_capitalization = true;
+
+    // Sadly enough, we can't mutate a string when iterating over its chars, likely because it could
+    // change the size of a char, "breaking" the char indices.
+    Ok(s.chars().fold(String::new(), |mut acc, c| {
+        if c.is_whitespace() {
+            acc.push(c);
+            need_capitalization = true;
+        } else if need_capitalization {
+            if !c.is_uppercase() {
```
Comment (djc) [inline askama/src/filters/mod.rs:528]: I think the best formulation of this would use `match` instead of `if` and `push_str()` or `extend()` instead of the loop.

Comment (GuillaumeGomez) [reply askama/src/filters/mod.rs]: :+1:

**Comment by GuillaumeGomez** [comment]
Applied suggestions and also added the `string` filter.

**Comment by djc** [comment]
What's the point of the `string` filter if simply using an expression already relies on `Display`?

**Comment by GuillaumeGomez** [comment]
In case you want to do operation on the generated string?

**Comment by djc** [comment]
Do you have an actual use case for that?

File: askama/src/filters/mod.rs [hunk 2]
```diff
@@ -502,7 +502,7 @@ pub fn center(src: impl ToString, dst_len: usize) -> Result<impl fmt::Display, I
     center(src.to_string(), dst_len)
 }
 
-/// Count the words in that string
+/// Counts the words in that string
```
Comment (djc) [inline askama/src/filters/mod.rs]: We're already inconsistent on the verb form and this is unrelated to the changes in this commit, so maybe just leave it for now?

(On the whole I think I prefer "Count"/"Center"/"Capitalize" over "Counts"/"Centers"/"Capitalizes".)

**Comment by GuillaumeGomez** [comment]
Not personally. My potential use case would be to format a struct that was turned into a string. But I could write my own filter for that. I can remove the second commit of the PR if you're not convinced. ;)

File: askama/src/filters/mod.rs [hunk 3]
```diff
@@ -511,6 +511,35 @@ pub fn wordcount(s: impl ToString) -> Result<usize, Infallible> {
     wordcount(s.to_string())
 }
 
+/// Returns a titlecased version of the value. I.e. words will start with uppercase letters, all
```
Comment (djc) [inline askama/src/filters/mod.rs:520]: Let's avoid "I.e." which is an obscure reference to latin words, and defer the second sentence to a second line of the docstring (after an empty line; and also update the book accordingly).

Also it appears "titlecase" is not an actual English word (per Merriam-Webster and DuckDuckGo), so just use "title case"? (Also in the docs.)

File: askama/src/filters/mod.rs [hunk 4]
```diff
@@ -511,6 +511,35 @@ pub fn wordcount(s: impl ToString) -> Result<usize, Infallible> {
...
+/// remaining characters are lowercase.
+#[inline]
+pub fn title(s: impl ToString) -> Result<String, Infallible> {
+    let s = s.to_string();
+    let mut need_capitalization = true;
+
+    // Sadly enough, we can't mutate a string when iterating over its chars, likely because it could
+    // change the size of a char, "breaking" the char indices.
+    Ok(s.chars().fold(String::new(), |mut acc, c| {
```
Comment (djc) [inline askama/src/filters/mod.rs:529]: Maybe `String::new()` -> `String::with_capacity(s.len())`?

File: book/src/filters.md [hunk 5]
```diff
@@ -20,25 +20,28 @@ Enable it with Cargo features (see below for more information).
...
+  * [`linebreaks`][#linebreaks],
+  * [`linebreaksbr`][#linebreaksbr],
+  * [`lower|lowercase`][#lower],
+  * [`safe`][#safe],
+  * [`string`][#string],
+  * [`title`][#title],
+  * [`trim`][#trim],
+  * [`truncate`][#truncate],
+  * [`upper|uppercase`][#upper],
+  * [`wordcount`][#wordcount]
```
Comment (djc) [inline book/src/filters.md:44]: Would prefer the other additions to this be in a separate commit.

File: askama/src/filters/mod.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply askama/src/filters/mod.rs:529]: Very good point!

File: book/src/filters.md (hunk 5 above)
Comment (GuillaumeGomez) [reply book/src/filters.md]: Done!

File: askama/src/filters/mod.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply askama/src/filters/mod.rs]: Reverted this change.

File: askama/src/filters/mod.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply askama/src/filters/mod.rs]: Fixed.

**Comment by djc** [comment]
> Not personally. My potential use case would be to format a struct that was turned into a string. But I could write my own filter for that. I can remove the second commit of the PR if you're not convinced. ;)

Yeah, let's wait until someone comes up with a use case.

**Comment by GuillaumeGomez** [comment]
Removed it!

File: askama/src/filters/mod.rs [hunk 6]
```diff
@@ -511,6 +511,42 @@ pub fn wordcount(s: impl ToString) -> Result<usize, Infallible> {
...
+/// Return a title cased version of the value. Words will start with uppercase letters, all
+/// remaining characters are lowercase.
+#[inline]
+pub fn title(s: impl ToString) -> Result<String, Infallible> {
+    let s = s.to_string();
+    let mut need_capitalization = true;
+
+    // Sadly enough, we can't mutate a string when iterating over its chars, likely because it could
+    // change the size of a char, "breaking" the char indices.
+    Ok(s.chars()
```
Comment (djc) [inline askama/src/filters/mod.rs:529]: I've been thinking about this but this latest reformatting (due to `new()` -> `with_capacity()`) I think pushes it over the line: a `for` loop would be a nice simplification instead of the `fold()` call.

File: askama/src/filters/mod.rs (hunk 3 above)
Comment (djc) [reply askama/src/filters/mod.rs]: Note:

> (after an empty line; and also update the book accordingly).

Comment (GuillaumeGomez) [reply askama/src/filters/mod.rs]: :sweat_smile:

File: askama/src/filters/mod.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply askama/src/filters/mod.rs:529]: Ok, updating.

**Comment by GuillaumeGomez** [comment]
Updated the book and replaced the `fold` iterator with a `for` loop.

**Review by djc — APPROVED** [review]
